[#3454] Replace CKEditor with Prosemirror

pi-sigma opened the pull request:
## Summary

+ Replaced CKEditor rich text fields and textfields with CKEditorWidget with
  ProsemirrorModelField.

+ Updated templatetags + tests for new ProsemirrorModelField

+ Added serializers for ProsemirrorFieldDocument in Elasticsearch indexing

  The Product.content and Category.description fields were converted to
  ProsemirrorModelField, but the Elasticsearch document indexing was trying
  to serialize the ProsemirrorFieldDocument objects directly, causing
  SerializationError.

  Added prepare_content() and prepare_categories() methods to convert
  ProsemirrorFieldDocument objects to HTML strings before indexing.

+ Modules affected:
  - configurations
  - cms
  - ssd
  - questionnaire
  - openklant
  - openzaak
  - pdc

## Issue References

Taiga user story: https://taiga.maykinmedia.nl/project/open-inwoner/us/3450?milestone=548
Taiga task: https://taiga.maykinmedia.nl/project/open-inwoner/task/3454

pi-sigma committed:
feat: replace CKEditor with Prosemirror

+ Replaced CKEditor rich text fields and textfields with CKEditorWidget with
  ProsemirrorModelField.

+ Updated templatetags + tests for new ProsemirrorModelField

+ Added serializers for ProsemirrorFieldDocument in Elasticsearch indexing

  The Product.content and Category.description fields were converted to
  ProsemirrorModelField, but the Elasticsearch document indexing was trying
  to serialize the ProsemirrorFieldDocument objects directly, causing
  SerializationError.

  Added prepare_content() and prepare_categories() methods to convert
  ProsemirrorFieldDocument objects to HTML strings before indexing.

Modules affected:
  - configurations
  - cms
  - ssd
  - questionnaire
  - openklant
  - openzaak
  - pdc

  Refs: https://taiga.maykinmedia.nl/project/open-inwoner/task/3454

pi-sigma committed:
refactor: delete unused admin classes

swrichards commented on src/open_inwoner/ssd/migrations/0007_ssdconfig_textfields.py:
```diff
@@ -0,0 +1,232 @@
+        r"\[([^\]]+)\]\(([^)]+)\)",  # Match [link_text](link)
+        r'<a href="\2" title="\1">\1</a>',  # Replace with <a href="link" title="link_text">link_text</a>
+        res,
+    )
```
Note we have `markdown==3.3.6` in our requirements, you could use that to do the conversion.

swrichards reviewed (commented):
Looks good! I wonder if it would be useful to add a `.safe_html` shorthand in django prosemirror to make the templates a bit easier to use.

I think you still need to add the dependency.

pi-sigma committed:
fix: add trailing newlines to compiled i18n files per editorconfig

swrichards commented on src/open_inwoner/configurations/migrations/0086_alter_siteconfiguration_warning_banner_text.py:
```diff
@@ -0,0 +1,97 @@
+        config.save()
+    except (IndexError, ValueError) as e:
+        # Skip records with problematic HTML content
+        print(f"Warning: Could not convert {field_name} for {model_name}: {e}")
```
Maybe a `logger.warning` would be more appropriate here.

swrichards commented on src/open_inwoner/pdc/migrations/0069_product_content.py:
```diff
@@ -0,0 +1,168 @@
+                src = attrs.get("src", "")
+
+                # Look up image by URL to get the correct ID
+                if "media" in src.lower() and "imageId" not in attrs:
```
I wonder if we need the `media` check, feels possible fragile -- can we use `MEDIA_PATH` and only use `Image.objects.get`? We handle the non-existing case anyway, don't think we have to limit this conditional further beyond checking for `imageId`.

swrichards commented on src/open_inwoner/utils/html.py:
```diff
@@ -39,14 +39,26 @@ def convert_first_row_to_th(html_tables):
+        return ""
+
+    # Check if this is a ProseMirror field
+    if hasattr(content, "doc"):
```
I think you can be more explicit here: I believe the fields will return a `ProsemirrorFieldDescriptor` instance, which you can import from `django_prosemirror.fields` for `isinstance`

pi-sigma replied:
The `isinstance` test fails for both `ProsemirrorFieldDescriptor` and `ProsemirrorModelField`. Not sure if this is due to a mismatch between test environment and the real world.

pi-sigma committed:
chore: upgrade django-prosemirror to latest

pi-sigma committed:
chore: remove CKEditor

Note: django-ckeditor is a dependency of mail-editor an remains
      (for now) in the project

Refs: https://taiga.maykinmedia.nl/project/open-inwoner/task/3454

pi-sigma committed:
feat: create Prosemirror-based TextPlugin to replace djangocms_text_ckeditor.TextPlugin

swrichards commented on src/open_inwoner/components/templates/components/Faq/Faq.html:
```diff
@@ -14,7 +14,7 @@ <h3 class="utrecht-heading-3">
                     {% link bold=True href='#'|add:answer_id icon='keyboard_arrow_down' secondary=True text=question.question toggle="open"  %}
                 </h3>
-                <div id="{{ answer_id }}" class="faq__answer">{{ question.answer|ckeditor_content|safe }}</div>
+                <div id="{{ answer_id }}" class="faq__answer">{{ question.answer|safe }}</div>
```
No `.html`?

swrichards commented on src/open_inwoner/pdc/models/faq.py:
```diff
@@ -24,7 +26,18 @@ class Question(OrderedModel):
-    answer = models.TextField(verbose_name=_("Antwoord"))
+    answer = ProsemirrorModelField(
+        _("Antwoord"),
+        allowed_node_types=[NodeType.PARAGRAPH],
```
It seems this field sometimes uses blockquote:

<img width="1673" height="330" alt="Image" src="https://github.com/user-attachments/assets/922d609e-0ad8-467d-90a6-f6a0756392e5" />

```suggestion
        _("Antwoord"),
        allowed_node_types=[NodeType.PARAGRAPH, NodeType.BLOCKQUOTE],
```

swrichards commented on src/open_inwoner/cms/plugins/models/text.py:
```diff
@@ -0,0 +1,14 @@
+    )
+
+    def __str__(self):
+        return str(self.body)[:50] if self.body else ""
```
Guess we might as well do `self.body.html` here, otherwise it's just a bunch of JSON.

swrichards commented on src/open_inwoner/pdc/utils.py:
```diff
@@ -6,19 +6,27 @@
-    md = markdown.Markdown()
-    html_string = md.convert(content)
+    # Check if content is a string
+    if isinstance(content, str):
+        # Legacy markdown conversion for old TextField content
+        md = markdown.Markdown()
+        html_string = md.convert(content)
+    else:
+        # ProsemirrorModelField - use .html property
+        html_string = content.html
```
Just want to double-check we have no better way of doing check than looking for the `.html` property (I know you tried, no worries if not, just jumped out at me again).

pi-sigma replied:
Did you mean some other part of the code? Here I'm using the `.html` property to assign the content, not to check for something.

swrichards replied:
Sorry, I mean checking if `content` is an instance of a prosemirror field. I was wondering whether we could check that directly, rather than inferring it from the presence of the `.html` attribute.

swrichards commented on src/open_inwoner/cms/footer/migrations/0003_rename_content_cmsflatpagemodel_content_tmp.py:
```diff
@@ -0,0 +1,101 @@
+        # If content doesn't look like HTML, wrap it in paragraphs
+        content = content.strip()
+        if not content.startswith("<"):
+            content = f"<p>{content}</p>"
```
What happens if we don't do this?

pi-sigma replied:
Nothing bad should happen, `html_to_doc` handles plain text correctly. This was defensive programming and should help to make the intent clear: make sure we're feeding HTML to django-prosemirror's parser. It's also perhaps a bit faster if the parser doesn't have to infer the structure. I could remove it though.

swrichards replied:
It's fine, defensive is good here.

swrichards commented on src/open_inwoner/configurations/migrations/0086_alter_siteconfiguration_warning_banner_text.py:
Two general comments:

1. Could we extract the conversion to a `migration_utils` helper, and re-use it? (Injecting the dependencies for the specific field)
2. I would still be inclined to merge these migrations if we can: I'd prefer they either all succeed or all fail, given how little we know about the state of the targets where this will be running.

pi-sigma replied:
~~Re 2: I found the issue. The migrations created automatically by Django have `history=True` for the addition of `ProsemirrorModelField`, which creates database triggers to track changes to the field. Postgres thenblocks the removal of the field in the subsequent migration because the triggers from the first field haven't been committed. The options are to do `atomic=False` for the addition of `ProsemirrorModelField`. which defeats the purpose of grouping them together, or `history=False`. I think we don't need `history=True`, which saves historical copies of the field; it's a singleton model used for configuration and we audit the change ourselves in the review. What do you think?~~

Disregard, the explanation is wrong and the change only helped accidentally

swrichards reviewed (approved):
Great job on this, couldn't really find any major issues except for the missing blockquote, and one or two questions/nitpicks, but otherwise it looks good to me.

pi-sigma committed:
chore: update changelog

swrichards commented on src/open_inwoner/pdc/migrations/0069_category_description_data.py:
Why can't this file use the helper?

pi-sigma replied:
Oversight or lost in rebase; fixed

swrichards commented on src/open_inwoner/pdc/tests/test_logging.py:
```diff
@@ -27,54 +27,78 @@ def setUp(self):
+        # Create product and change it to generate a log entry
+        from django.contrib.admin.models import CHANGE, LogEntry
+        from django.contrib.contenttypes.models import ContentType
+
```
Why the inline imports?

pi-sigma replied:
Forgot to move them after experimenting

swrichards commented on src/open_inwoner/utils/migration_operations.py:
```diff
@@ -1,59 +1,337 @@
+    # Remove empty tags iteratively (nested empty tags need multiple passes)
+    # Matches: <tag></tag> or <tag> </tag> or <tag attrs></tag>
+    prev = None
+    max_iterations = 10  # Safety limit
```
What kind of limits do we run up against? If we do reach our limit, if I understand this correctly, we'll still have to deal with the `IndexError`s later. We're not recursing, it might just take a long time. If the fear is infinite loop, I would set the limit much higher so we do eventually time out if we reach an infinite edge case, but will hopefully cover all practical scenarios.

pi-sigma replied:
Meant as a defense against regex edge cases or malformed HTML that would get the function stuck in a loop. Shouldn't happen but well... I've set it to 50. Seems enough? Real-world HTML shouldn't contain that many levels of tag nesting (though who knows).

swrichards reviewed (approved):
Good work here, turns out (as ever) there were more edge cases than we realized. I think we've done all we reasonable can to smooth the migration.

Just a few questions/nitpicks, can do a quick follow up review and merge before tomorrow morning's meeting.

swrichards commented on src/open_inwoner/utils/migration_operations.py:
```diff
@@ -1,59 +1,339 @@
+        try:
+            # Convert markdown to HTML if needed
+            if convert_markdown:
+                content = markdown.markdown(content, extensions=["extra"])
```
Final thought on this: markdown and HTML are mutually exclusive right? And if so, is there a chance the earlier cleaning operations will generate errors (because they're working on markdown, not HTML), or does thew built-in HTML parser silently ignore everything in that case?

pi-sigma replied:
Good catch! They're not exclusive (markdown can contain inline HTML) and the cleaning operations work on HTML not markdown, but there could be bugs: e.g. if the cleaning operations modify/strip inline HTML and the markdown conversion is feeded corrupted markdown. I'll fix it, thanks.

swrichards reviewed (approved):
Epic PR, I think we're there. Just one final question, probably not an issue but just wanted to make sure.
